From e782332347f352b028d05a2f0d96420db15dfc31 Mon Sep 17 00:00:00 2001
From: Jostein Austvik Jacobsen <josteinaj@gmail.com>
Date: Thu, 29 Jan 2015 15:28:30 +0100
Subject: [PATCH] added xpl-href and xpl-line to profile elements

---
 src/com/xmlcalabash/core/XProcRuntime.java | 2 ++
 1 file changed, 2 insertions(+)

diff --git src/com/xmlcalabash/core/XProcRuntime.java src/com/xmlcalabash/core/XProcRuntime.java
index fadb83a..0a8e0be 100644
--- src/com/xmlcalabash/core/XProcRuntime.java
+++ src/com/xmlcalabash/core/XProcRuntime.java
@@ -963,6 +963,8 @@ public class XProcRuntime {
         String name = step.getType().getClarkName();
         profileWriter.addAttribute(profileType, name);
         profileWriter.addAttribute(profileName, step.getStep().getName());
+        profileWriter.addAttribute(new QName("", "xpl-href"), ""+step.getStep().xplFile());
+        profileWriter.addAttribute(new QName("", "xpl-line"), ""+step.getStep().xplLine());
         profileWriter.startContent();
     }
 
-- 
2.1.0

