Index: cz/vutbr/web/domassign/DeclarationTransformer.java
===================================================================
--- cz/vutbr/web/domassign/DeclarationTransformer.java
+++ cz/vutbr/web/domassign/DeclarationTransformer.java
@@ -110,14 +110,14 @@
  */
 public class DeclarationTransformer {
 
-	private static Logger log = LoggerFactory
+	protected static Logger log = LoggerFactory
 			.getLogger(DeclarationTransformer.class);
 
 	/**
 	 * Inherit acceptance flags
 	 */
-	private static final boolean AVOID_INH = true;
-	private static final boolean ALLOW_INH = false;
+	protected static final boolean AVOID_INH = true;
+	protected static final boolean ALLOW_INH = false;
 
 	/**
 	 * Cache of parsing methods
@@ -129,7 +129,7 @@
 	 */
 	private static final DeclarationTransformer instance;
 
-	private static final TermFactory tf = CSSFactory.getTermFactory();
+	protected static final TermFactory tf = CSSFactory.getTermFactory();
 	private static final SupportedCSS css = CSSFactory.getSupportedCSS();
 
 	static {
@@ -225,7 +225,7 @@
 	/**
 	 * Sole constructor
 	 */
-	private DeclarationTransformer() {
+	protected DeclarationTransformer() {
 		this.methods = parsingMethods();
 	}
 
