Index: cz/vutbr/web/csskit/ElementUtil.java
===================================================================
--- cz/vutbr/web/csskit/ElementUtil.java	(revision 656)
+++ cz/vutbr/web/csskit/ElementUtil.java	(working copy)
@@ -88,7 +88,7 @@
          			attributeValue = " " + attributeValue + " ";
          			return attributeValue.matches(".* " + value + " .*");
          		case DASHMATCH:
-         			return attributeValue.matches("^" + value + "(\\|.*)*");
+         			return attributeValue.matches("^" + value + "(-.*|$)");
          		default:
          			return true;
      		}
