Index: cz/vutbr/web/css/CSSFactory.java
===================================================================
--- cz/vutbr/web/css/CSSFactory.java
+++ cz/vutbr/web/css/CSSFactory.java
@@ -6,6 +6,8 @@
 import java.lang.reflect.Method;
 import java.net.MalformedURLException;
 import java.net.URL;
+import javax.xml.transform.Source;
+import javax.xml.transform.URIResolver;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -68,6 +70,8 @@
 
 	private static Class<? extends NodeData> ndImpl;
 
+	private static URIResolver ur;
+
 	/**
 	 * Registers new TermFactory instance
 	 * 
@@ -221,6 +225,32 @@
 	}
 
 	/**
+	 * Registers URIResolver.
+	 * 
+	 * @param uriResolver
+	 *            New URIResolver instance
+	 */
+	public static final void registerURIResolver(URIResolver newURIResolver) {
+		ur = newURIResolver;
+	}
+
+	/**
+	 * Returns registered URIResolver.
+	 * 
+	 * @return URIResolver instance
+	 */
+	public static final URIResolver getURIResolver() {
+		if (ur == null) {
+			ur = new URIResolver() {
+				public Source resolve(String href, String base) {
+					return null;
+				}
+			};
+		}
+		return ur;
+	}
+
+	/**
 	 * Parses URL into StyleSheet
 	 * 
 	 * @param url
Index: cz/vutbr/web/csskit/antlr/CSSInputStream.java
===================================================================
--- cz/vutbr/web/csskit/antlr/CSSInputStream.java
+++ cz/vutbr/web/csskit/antlr/CSSInputStream.java
@@ -8,15 +8,22 @@
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.InputStreamReader;
+import java.net.MalformedURLException;
 import java.net.URL;
 import java.net.URLConnection;
 import java.nio.charset.Charset;
 import java.util.zip.GZIPInputStream;
 
+import javax.xml.transform.Source;
+import javax.xml.transform.TransformerException;
+import javax.xml.transform.URIResolver;
+
 import org.antlr.runtime.ANTLRInputStream;
 import org.antlr.runtime.ANTLRReaderStream;
 import org.antlr.runtime.CharStream;
 
+import cz.vutbr.web.css.CSSFactory;
+
 /**
  * Wraps ANTLR stream with useful decorations,
  * mainly to allow switching encoding on lexer
@@ -45,6 +52,7 @@
 	 */
 	private String encoding;
 	
+	private static final URIResolver ur = CSSFactory.getURIResolver();
 	
 	public static CSSInputStream stringStream(String source) throws IOException {
 		CSSInputStream stream = new CSSInputStream();
@@ -62,6 +70,14 @@
 	public static CSSInputStream urlStream(URL source) throws IOException {
 		CSSInputStream stream = new CSSInputStream();
 		
+		try {
+			Source resolvedSource = ur.resolve(source.toString(), "");	
+			if (resolvedSource != null)
+				source = new URL(resolvedSource.getSystemId());
+		} catch (TransformerException e) {
+		} catch (MalformedURLException e) {
+		}
+
 		stream.base = source;
 		stream.encoding = Charset.defaultCharset().name();
 		
@@ -201,6 +217,13 @@
 	}
 	
 	public void setBase(URL base) {
+		try {
+			Source resolvedBase = ur.resolve(base.toString(), "");	
+			if (resolvedBase != null)
+				base = new URL(resolvedBase.getSystemId());
+		} catch (TransformerException e) {
+		} catch (MalformedURLException e) {
+		}
 	    this.base = base;
 	}
 	
